### ISSUE TITLE
Issue: Custom class as Key

### ISSUE THREAD

**Issue by jasonlawless** [upscalebaby/generic-serializable-dictionary#10]
I'm not sure if this is strictly for string keys or if it was intended to work with any class, but I have a dictionary that uses a class I wrote as a Key, and I use it on a ScriptableObject. My class DOES have a <code>override ToString</code> as well as an <code>implicit operator MyClass(string key)</code> which converts it back from string. I did these so that my JSON Serializer would work correctly using this class as a Dictionary Key, and it does.

But the few issues I see using the GenericDictionary whilst using a class instead of a string key are:
1. In the inspector, the Key TextField is not showing a Label nor a TextField, it's a blank key.
2. When a key is added and between Play sessions the <code>TryGetValue</code> actually works and finds the correct key. BUT, when scripts recompile, the <code>TryGetValue</code> comes back false even though the Element is STILL in the dictionary on the inspector.

I hope this made sense. Keep in mind this was for a <code>ScriptableObject</code>. I think the issue lies somewhere in the Key not saving properly in between script compilations. But I don't know tbh.

PS> I am using <code>Enter Play Mode Settings</code> with <b>Reload Domain</b> switched off.

**Comment by upscalebaby** [comment]
> I'm not sure if this is strictly for string keys or if it was intended to work with any class

Yes it's supposed to work for all use cases, including custom serializable classes as Keys (regardless if it's on a MonoBehaviour or ScriptableObject) so I'm glad you opened an issue on this so we hopefully can resolve it!

First and foremost are you using the 2020.1 release or are you working from the master that supports 2020.2? Also what Unity version are you working on?

I just tried this on Unity 2020.2.5f1 using a custom class as a Key on a ScriptableObject and it appears to work:

![image](https://user-images.githubusercontent.com/12954999/108778834-47ffea00-7566-11eb-8f0b-168750cd993e.png)

This is the source for the Key class:

`using UnityEngine;

[System.Serializable]
public class CustomClass
{
    [TextArea]
    public string textArea;
    public GameObject someObject;
    public float someFloat;
}
`

Regarding this:

> BUT, when scripts recompile, the TryGetValue comes back false even though the Element is STILL in the dictionary on the inspector.

Could you verify if this is fixed by turning off all enter play mode settings to ensure everything is reloaded?

**Comment by jasonlawless** [comment]
I am so glad that it is supposed to work for all use cases! And thank you for your prompt reply as well!
To answer your question, yes I am using <code>GenericDictionary</code> <i>2020.1</i>
My current Unity version is: <i>2020.1.17f</i>

Here is my source for the class I am trying to use:
<code>
using System;
using UnityEngine;
using Newtonsoft.Json;

[JsonObject][Serializable]
public struct ChunkID : IEquatable<ChunkID>
{
    public static ChunkID FromWorldPos(int x, int y, int z) => new ChunkID(x >> 4, y >> 4, z >> 4);
    public static ChunkID FromWorldPos(Vector3 worldPosition) =>
        FromWorldPos((int)worldPosition.x, (int)worldPosition.y, (int)worldPosition.z);

    public readonly int X;
    public readonly int Y;
    public readonly int Z;

    public ChunkID(int x, int y, int z)
    {
        X = x; Y = y; Z = z;
    }
    public ChunkID(Vector3 worldPosition) : 
        this((int)worldPosition.x, (int)worldPosition.y, (int)worldPosition.z) { }


    #region EQUATABLE INTERFACE

    public bool Equals(ChunkID other) => X == other.X && Y == other.Y && Z == other.Z;

    public override bool Equals(object obj)
    {
        if (obj is null) return false;
        return obj is ChunkID other && Equals(other);
    }

    public override int GetHashCode()
    {
        unchecked
        {
            int hashCode = X;
            hashCode = (hashCode * 397) ^ Y;
            hashCode = (hashCode * 397) ^ Z;
            return hashCode;
        }
    }

    public static bool operator ==(ChunkID left, ChunkID right) => left.Equals(right);
    public static bool operator !=(ChunkID left, ChunkID right) => !left.Equals(right);


    #endregion EQUATABLE INTERFACE

    public override string ToString()
    {
        return $"{X},{Y},{Z}";
    }

    public static implicit operator ChunkID(string key)
    {
        string[] keys = key.Split(',');
        int x = int.Parse(keys[0]);
        int y = int.Parse(keys[1]);
        int z = int.Parse(keys[2]);
        return new ChunkID(x, y, z);
    }
}
</code>

Maybe you can see something wrong with my class and give me some pointers? There's a good chance I'm just being a n00b right now lol.
Also, I have turned off <b>Play Mode Settings</b> and the result is the same as if I was re-compiling script before entering Play mode. In other words, it doesn't save the key data in between play sessions.

**Comment by upscalebaby** [comment]
Okay I _think_ I know what's causing the first problem. I downloaded 2020.1.17f1 and think I've solved the first issue. Could you verify by replacing the GenericDictionaryPropertyDrawer inside the Editor folder with this one: https://github.com/upscalebaby/generic-serializable-dictionary/blob/master/Assets/Editor/GenericDictionaryPropertyDrawer.cs

Does that affect any of the issues?

**Comment by jasonlawless** [comment]
I tried it with the code you provided, and the result is the same. Keys don't show and it doesn't find existing key.

![image](https://user-images.githubusercontent.com/25869800/108785647-b3c26280-7526-11eb-9846-71400ed43685.png)

**Comment by jasonlawless** [comment]
Yes! That's the issue! All the public fields were `readonly` and that is what was causing these issues. I removed the `readonly` keyword and now it finds the key properly and also display it correctly!

![image](https://user-images.githubusercontent.com/25869800/108786179-e3be3580-7527-11eb-8e77-06e884993f55.png)

Do you think this is something that is easy to fix? Or am I just better off keeping those fields away from `readonly`? I would much prefer to keep `readonly` to prevent any accidental changes because it can cause a chain reaction of orphaned data.

**Comment by upscalebaby** [comment]
Great! I was just about to say readonly fields doesn't display natively in the inspector and so doesn't do so here either - I do know there are workarounds for that specific problem that _should_ carry over fine when used with this dictionary.

Regarding PropertyDrawers they should work perfectly with the new PropertyDrawer I linked, I tried it with the old 2020.1 version and noticed it does work but the custom PropetyDrawers height isn't respected. Overall the new PropertyDrawer is much more robust and the price of not looking quite as pretty - I'm currently looking into ways to get the best of both worlds and will update the repo once I've landed on something.

So to sum up you only need to create a custom property drawer for your class and your good to go. Unity will use your custom drawers for displaying everything in the dictionary (both Key/Value).

Hope that helps!

**Comment by jasonlawless** [comment]
Yes it does! Thank you for your time and help! I very much appreciate it man. Keep it up!
